torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -580,6 +580,65 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
   }
 };
 
+class AsyncAllreduceCoalescedWork: public ProcessGroupGloo::AsyncWork {
+ public:
+  AsyncAllreduceCoalescedWork(
+    std::shared_ptr<gloo::Context> context,
+    std::vector<at::Tensor>& inputs,
+    ReduceOp reduceOp,
+    uint32_t tag)
+    : context(std::move(context)), inputs(inputs), reduceOp(reduceOp), tag(tag) {}
+
+  std::shared_ptr<gloo::Context> context;
+  std::vector<at::Tensor> inputs;
+  const ReduceOp reduceOp;
+  const uint32_t tag;
+
+  void run() override {
+    allreduceCoalesced(inputs);
+  }
+
+ private:
+  template <typename T>
+  void getFunction(gloo::AllreduceOptions::Func& fn, const ReduceOp op) {
+    fn = toFunction<T>(op);
+  }
+
+  gloo::AllreduceOptions::Func getFunction(
+      const at::ScalarType& dtype,
+      const ReduceOp op) {
+    gloo::AllreduceOptions::Func fn;
+    GENERATE_ALL_TYPES(dtype, getFunction, fn, op);
+    return fn;
+  }
+
+  void allreduce(at::Tensor& tensor) {
```

Comment:
could it still be vector<at::Tensor> and input size will be 1? in this way, you do not need to overwrite the allreduce() function but just inherit from AsyncAllReduceWork

Reply:
oh good idea, changed.